### PR TITLE
Expose nint/nuint overloads for BitOperations

### PR DESCRIPTION
**Pull request by deeprobin**
## Description

Hello everybody, in this pull request I am implementing the proposed API to #54390 by @tannergooding.

Feel free to leave your feedback :).

## Proposed API

> 
> ```cs
> namespace System.Numerics
> {
>     public static partial class BitOperations
>     {
>         public static bool IsPow2(nint value);
>         public static bool IsPow2(nuint value);
> 
>         public static nuint RoundUpToPowerOf2(nuint value);
> 
>         public static int LeadingZeroCount(nuint value);
> 
>         public static int Log2(nuint value);
> 
>         public static int PopCount(nuint value);
> 
>         public static int TrailingZeroCount(nint value);
>         public static int TrailingZeroCount(nuint value);
> 
>         public static nuint RotateLeft(nuint value, int offset);
>         public static nuint RotateRight(nuint value, int offset);
>     }
> }
> ```
> Proposal of @tannergooding

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=58733) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=58733) before we can accept your contribution.<br/><br/>:x: deeprobin [sign now](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=58733)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/runtime?pullRequest=58733) it.</sub>

File: src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs [hunk 1]
```diff
@@ -145,6 +146,26 @@ public static ulong RoundUpToPowerOf2(ulong value)
...
+        /// <param name="value">The value.</param>
+        /// <returns>
+        /// The smallest power of 2 which is greater than or equal to <paramref name="value"/>.
+        /// If <paramref name="value"/> is 0 or the result overflows, returns 0.
+        /// </returns>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        [CLSCompliant(false)]
+        public static nuint RoundUpToPowerOf2(nuint value)
+        {
+            if (Environment.Is64BitProcess)
```
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs:161]: In situations like this one, we often use `#if TARGET_64BIT` in CoreLib to make the IL smaller and easier for JIT to inline.

File: src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs [hunk 2]
```diff
@@ -105,7 +107,7 @@ public static uint RoundUpToPowerOf2(uint value)
 #endif
             }
 
-            // Based on https://graphics.stanford.edu/~seander/bithacks.html#RoundUpPowerOf2
+            // Based on
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs:110]: Why did you remove the link?

Comment (deeprobin) [reply src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs]: Hi @stephentoub,

thanks for bringing this to my attention. Have removed this by mistake. I have put the comment back into `BitOperations.cs`.

Best regards
Robin

File: src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs (hunk 1 above)
Comment (deeprobin) [reply src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs]: Hello @jkotas,

have now replaced the check with the preprocessor variable, which hopefully inlines the JIT better.

Best regards
Robin

File: src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs [hunk 3]
```diff
@@ -218,6 +238,23 @@ public static int LeadingZeroCount(ulong value)
...
+        /// <summary>
+        /// Count the number of leading zero bits in a mask.
+        /// Similar in behavior to the x86 instruction LZCNT.
+        /// </summary>
+        /// <param name="value">The value.</param>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        [CLSCompliant(false)]
+        public static int LeadingZeroCount(nuint value)
+        {
+            if (Environment.Is64BitProcess)
```
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs:250]: Use `TARGET_64BIT`

File: src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs [hunk 4]
```diff
@@ -432,6 +485,22 @@ static int SoftwareFallback(ulong value)
...
+        /// Returns the population count (number of bits set) of a mask.
+        /// Similar in behavior to the x86 instruction POPCNT.
+        /// </summary>
+        /// <param name="value">The value.</param>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        [CLSCompliant(false)]
+        public static int PopCount(nuint value)
+        {
+#if TARGET_64BIT
+                return PopCount((ulong)value);
```
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs:498]: Nit: Indentation

File: src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs [hunk 5]
```diff
@@ -76,14 +76,16 @@ public static class BitOperations
         /// </summary>
         /// <param name="value">The value.</param>
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        internal static bool IsPow2(nint value) => (value & (value - 1)) == 0 && value > 0;
+        [CLSCompliant(false)]
```
Comment (tannergooding) [inline src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs:79]: This shouldn't need `CLSCompliant(false)`, `IntPtr` itself is compliant.

File: src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs (hunk 3 above)
Comment (tannergooding) [inline src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs:250]: Could you do the same `#if TARGET_64BIT` to all of these `if (Enviroonment.Is64BitProcess)` paths?

**Comment by nelsondani** [comment]
[Uploading provider-relief-fund-nursing-home-quality-incentive-payment-allocations.pdf…]()

File: src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs (hunk 3 above)
Comment (deeprobin) [reply src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs]: Done

File: src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs (hunk 4 above)
Comment (deeprobin) [reply src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs]: Done

File: src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs (hunk 5 above)
Comment (deeprobin) [reply src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs:79]: Uhm the build fails without this attribute :neutral_face:.
Error: CS3001

```
E:\dotnet\runtime>build.cmd
  Wiederherzustellende Projekte werden ermittelt...
  Das Tool "coverlet.console" (Version 1.7.2) wurde wiederhergestellt. Verfügbare Befehle: coverlet
  Das Tool "dotnet-reportgenerator-globaltool" (Version 4.5.8) wurde wiederhergestellt. Verfügbare Befehle: reportgenerator
  Das Tool "microsoft.dotnet.xharness.cli" (Version 1.0.0-prerelease.21451.1) wurde wiederhergestellt. Verfügbare Befehle: xharness
  Das Tool "microsoft.visualstudio.slngen.tool" (Version 6.1.10) wurde wiederhergestellt. Verfügbare Befehle: slngen

  Die Wiederherstellung war erfolgreich.
  Alle Projekte sind für die Wiederherstellung auf dem neuesten Stand.
  Wiederherzustellende Projekte werden ermittelt...
  Alle Projekte sind für die Wiederherstellung auf dem neuesten Stand.
  System.Private.CoreLib.Generators -> E:\dotnet\runtime\artifacts\bin\System.Private.CoreLib.Generators\netstandard2.0-Debug\System.Private.CoreLib.Generators.dll
E:\dotnet\runtime\src\libraries\System.Private.CoreLib\src\System\Numerics\BitOperations.cs(87,41): error CS3001: Argumenttyp "nuint" ist nicht CLS-kompatibel. [E:\dotnet\runtime\src\coreclr\System.Private.CoreLib\System.Private.CoreLib.csproj]

Fehler beim Buildvorgang.

E:\dotnet\runtime\src\libraries\System.Private.CoreLib\src\System\Numerics\BitOperations.cs(87,41): error CS3001: Argumenttyp "nuint" ist nicht CLS-kompatibel. [E:\dotnet\runtime\src\coreclr\System.Private.CoreLib\System.Private.CoreLib.csproj]
    0 Warnung(en)
    1 Fehler

Verstrichene Zeit 00:02:22.78
Build failed with exit code 1. Check errors above.
Some builds failed:
        Configuration: Debug, Architecture: x64
```

File: src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs (hunk 3 above)
Comment (deeprobin) [reply src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs]: Done

File: src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs (hunk 5 above)
Comment (tannergooding) [reply src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs:79]: That's complaining about L87 (the one taking `nuint`) not about L79 (the one taking `nint`).

Comment (deeprobin) [reply src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs]: Thank you. I assumed that UIntPtr and IntPtr interact basically the same. 
Now the build is running for me.

File: src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs [hunk 6]
```diff
@@ -551,6 +643,26 @@ public static uint RotateLeft(uint value, int offset)
         public static ulong RotateLeft(ulong value, int offset)
             => (value << offset) | (value >> (64 - offset));
 
+        /// <summary>
+        /// Rotates the specified value left by the specified number of bits.
+        /// Similar in behavior to the x86 instruction ROL.
+        /// </summary>
+        /// <param name="value">The value to rotate.</param>
+        /// <param name="offset">The number of bits to rotate by.
```
Comment (deeprobin) [inline src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs:651]: Would you improve the wording of the documentation comment (I am not a native English speaker)?

File: src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs [hunk 7]
```diff
@@ -576,5 +688,25 @@ public static uint RotateRight(uint value, int offset)
...
         public static ulong RotateRight(ulong value, int offset)
             => (value >> offset) | (value << (64 - offset));
+
+        /// <summary>
+        /// Rotates the specified value right by the specified number of bits.
+        /// Similar in behavior to the x86 instruction ROR.
+        /// </summary>
+        /// <param name="value">The value to rotate.</param>
+        /// <param name="offset">The number of bits to rotate by.
+        /// Any value outside the range [0..31] is treated as congruent mod 32 on a 32-bit process,
```
Comment (deeprobin) [inline src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs:698]: Would you improve the wording of the documentation comment (I am not a native English speaker)?

**Comment by deeprobin** [comment]
Why does the Test Runner say that the methods do not exist?

**Comment by GrabYourPitchforks** [comment]
> Why does the Test Runner say that the methods do not exist?

@deeprobin Don't forget to update the ref assembly (https://raw.githubusercontent.com/dotnet/runtime/main/src/libraries/System.Runtime/ref/System.Runtime.cs). Easiest way to update it would probably be to look at how the other `BitOperations` API stubs are exposed, then copy + paste those as new lines, renaming methods and parameters as needed.

**Comment by danmoseley** [comment]
Better still to use the tool so there's no extra diffs:

https://github.com/dotnet/runtime/blob/main/docs/coding-guidelines/updating-ref-source.md

**Comment by deeprobin** [comment]
> 
> 
> Better still to use the tool so there's no extra diffs:
> 
> https://github.com/dotnet/runtime/blob/main/docs/coding-guidelines/updating-ref-source.md

@danmoseley
The tool does not seem to work for me :/.

```
E:\dotnet\runtime>cd src

E:\dotnet\runtime\src>dotnet msbuild /t:GenerateReferenceAssemblySource
Could not execute because the application was not found or a compatible .NET SDK is not installed.
Possible reasons for this include:
  * You intended to execute a .NET program:
      The application 'msbuild' does not exist.
  * You intended to execute a .NET SDK command:
      A compatible installed .NET SDK for global.json version [6.0.100-rc.1.21430.12] from [E:\dotnet\runtime\global.json] was not found.
      Install the [6.0.100-rc.1.21430.12] .NET SDK or update [E:\dotnet\runtime\global.json] with an installed .NET SDK:
        5.0.102 [C:\PROGRA~1\dotnet\sdk]
        6.0.100-preview.7.21379.14 [C:\PROGRA~1\dotnet\sdk]

E:\dotnet\runtime\src>
```

**Comment by deeprobin** [comment]
> 
> 
> > Why does the Test Runner say that the methods do not exist?
> 
> @deeprobin Don't forget to update the ref assembly (https://raw.githubusercontent.com/dotnet/runtime/main/src/libraries/System.Runtime/ref/System.Runtime.cs). Easiest way to update it would probably be to look at how the other `BitOperations` API stubs are exposed, then copy + paste those as new lines, renaming methods and parameters as needed.

@GrabYourPitchforks Thank you. Thank you, manually updated the ref assembly in the last commit.

File: src/libraries/System.Runtime.Extensions/tests/System/Numerics/BitOperationsTests.cs [hunk 8]
```diff
@@ -228,6 +324,90 @@ public static void BitOps_TrailingZeroCount_long(long n, int expected)
...
+        [InlineData(ushort.MaxValue << 16, 16)]
+        [InlineData(ushort.MaxValue << 19, 19)]
+        [InlineData(int.MaxValue << 5, 5)]
+        [InlineData(3 << 27, 27)]
+        [InlineData(int.MaxValue, 0)]
+        public static void BitOps_TrailingZeroCount_nint(int n, int expected)
+#endif
+        {
+            int actual = BitOperations.TrailingZeroCount((nint) n);
+            Assert.Equal(expected, actual);
```
Comment (deeprobin) [inline src/libraries/System.Runtime.Extensions/tests/System/Numerics/BitOperationsTests.cs:408]: Why is this 64?

> ```
>     System.Numerics.Tests.BitOperationsTests.BitOps_TrailingZeroCount_nint(n: 0, expected: 32) [FAIL]
>       Assert.Equal() Failure
>       Expected: 32
>       Actual:   64
>       Stack Trace:
>         /_/src/libraries/System.Runtime.Extensions/tests/System/Numerics/BitOperationsTests.cs(408,0): at System.Numerics.Tests.BitOperationsTests.BitOps_TrailingZeroCount_nint(Int32 n, Int32 expected)
> ```
> Test Runner Output: https://helixre8s23ayyeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-58733-merge-20cb53393b3c40eca0/System.Runtime.Extensions.Tests/1/console.578f5a72.log?sv=2019-07-07&se=2021-09-30T17%3A42%3A26Z&sr=c&sp=rl&sig=dcC1qEiBBMgOi9HLSjMNa1re1pNJvHbzcyQP1sHQ0XY%3D

File: src/libraries/System.Runtime.Extensions/tests/System/Numerics/BitOperationsTests.cs [hunk 9]
```diff
@@ -269,6 +449,31 @@ public static void BitOps_Log2_ulong(ulong n, int expected)
...
+        [InlineData(uint.MaxValue, 32 - 1)]
+#if TARGET_64BIT
+        [InlineData(ulong.MaxValue, 64 - 1)]
+        public static void BitOps_Log2_nuint(uint n, int expected)
+#else
+        public static void BitOps_Log2_nuint(ulong n, int expected)
+#endif
+        {
+            int actual = BitOperations.Log2((nuint) n);
+            Assert.Equal(expected, actual);
```
Comment (deeprobin) [inline src/libraries/System.Runtime.Extensions/tests/System/Numerics/BitOperationsTests.cs:474]: Why is this 64?

> ```
>     System.Numerics.Tests.BitOperationsTests.BitOps_TrailingZeroCount_nint(n: 0, expected: 32) [FAIL]
>       Assert.Equal() Failure
>       Expected: 32
>       Actual:   64
>       Stack Trace:
>         /_/src/libraries/System.Runtime.Extensions/tests/System/Numerics/BitOperationsTests.cs(408,0): at System.Numerics.Tests.BitOperationsTests.BitOps_TrailingZeroCount_nint(Int32 n, Int32 expected)
> ```
> Test Runner Output:     System.Numerics.Tests.BitOperationsTests.BitOps_TrailingZeroCount_nuint(n: 0, expected: 32) [FAIL]
      Assert.Equal() Failure
      Expected: 32
      Actual:   64
      Stack Trace:
        /_/src/libraries/System.Runtime.Extensions/tests/System/Numerics/BitOperationsTests.cs(366,0): at System.Numerics.Tests.BitOperationsTests.BitOps_TrailingZeroCount_nuint(UInt32 n, Int32 expected)

File: src/libraries/System.Runtime.Extensions/tests/System/Numerics/BitOperationsTests.cs [hunk 10]
```diff
@@ -228,6 +324,90 @@ public static void BitOps_TrailingZeroCount_long(long n, int expected)
...
+        [InlineData((uint)ushort.MaxValue << 16, 16)]
+        [InlineData((uint)ushort.MaxValue << 19, 19)]
+        [InlineData(uint.MaxValue << 5, 5)]
+        [InlineData(3u << 27, 27)]
+        [InlineData(uint.MaxValue, 0)]
+        public static void BitOps_TrailingZeroCount_nuint(uint n, int expected)
+#endif
+        {
+            int actual = BitOperations.TrailingZeroCount((nuint) n);
+            Assert.Equal(expected, actual);
```
Comment (deeprobin) [inline src/libraries/System.Runtime.Extensions/tests/System/Numerics/BitOperationsTests.cs:366]: Why is this 64?

> ```
>     System.Numerics.Tests.BitOperationsTests.BitOps_TrailingZeroCount_nuint(n: 0, expected: 32) [FAIL]
>       Assert.Equal() Failure
>       Expected: 32
>       Actual:   64
>       Stack Trace:
>         /_/src/libraries/System.Runtime.Extensions/tests/System/Numerics/BitOperationsTests.cs(366,0): at System.Numerics.Tests.BitOperationsTests.BitOps_TrailingZeroCount_nuint(UInt32 n, Int32 expected)
> ```

> Test Runner Output: https://helixre8s23ayyeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-58733-merge-20cb53393b3c40eca0/System.Runtime.Extensions.Tests/1/console.578f5a72.log?sv=2019-07-07&se=2021-09-30T17%3A42%3A26Z&sr=c&sp=rl&sig=dcC1qEiBBMgOi9HLSjMNa1re1pNJvHbzcyQP1sHQ0XY%3D

**Comment by danmoseley** [comment]
> The tool does not seem to work for me 

 Try using the dotnet at the root of the repo. It pulls down and uses the correct one. Alternatively it should fix it if you install the RC1 SDK

**Review by tannergooding — APPROVED** [review]

**Comment by tannergooding** [comment]
@pgovind could you give this a secondary review and merge if everything looks good?

**Review by pgovind — APPROVED** [review]
LGTM! Thanks for the work here @deeprobin !

**Comment by deeprobin** [comment]
I'm glad to hear that :).
